### DISCUSSION TITLE
Questions and Sugeestions

### DISCUSSION THREAD

**Discussion by ghost** [phil-opp/blog_os#930] [Questions] [answered]
Hi. I have some questions and suggestions about the project:
A) Questions:
- is this correct? If I runnig the system , it will print Hello World and infinite dots?
- would it be possible to create a small GUI similar at least to Windows 98?
B) Suggestions:
- I will be happy to help you translate the website into Polish
I don't know what else to add yet, but I'll think. Overall, I'd like to help

**Comment by ghost** [comment]
Oh yeah, I'd like to ask: How to add a readline function? How to make a filesystem for the system?
How to add access to std so that I can use the standard library functions

**Reply by phil-opp** [reply]
Could you clarify what you mean with a readline function? If you mean keyboard input, take a look at our [Hardware Interrupts](https://os.phil-opp.com/hardware-interrupts/) post.

>  How to make a filesystem for the system?

You need to create a disk driver to access the individual bytes on the disk. Then you can read/create the file system structure on it. I haven't tried that yet, so I can't help you with the details.



> How to add access to std so that I can use the standard library functions

The standard library functions are wrappers for operating system functionality. So you need to implement all the required features (e.g. files, threads, processes) for your OS yourself.

**Comment by phil-opp** [comment, marked as answer]
> If I runnig the system , it will print Hello World and infinite dots?

Yes. The dots are printed on the regular timer interrupts. Additionally, keyboard input should work.



> would it be possible to create a small GUI similar at least to Windows 98?

I'm currently working on a new edition of the blog, which will use a pixel-based framebuffer. This should make it possible to draw your own GUI.



> I will be happy to help you translate the website into Polish

Thanks for offering your help! We're always happy to add more translations to the blog.

**Comment by ghost** [comment]
Recently I wrote an assembly language bootloader using FASM (Flat Assembly) and now I have a question how to attach a rust kernel to this. According to this tutorial, the structure of the project should be:

Project files:
Bootloader 
- boot.asm
- loader.asm
- loader.c
Kernel
- main.c
- other files 
Because this tutorial was about C.
I would like all C language files to be Rust (so that it works),
I'd also like to configure Cargo.toml and build.rs well. But I don't know how since I am a beginner.

You can find my files here in the bootloader directory
Unfortunately, in this repository I did not give C language files. If you need them, let me know
(https://github.com/PowerAppsInc/BirdOS)

**Reply by phil-opp** [reply]
Try compiling your kernel as staticlib. You can do that by creating a `lib.rs` instead of a `main.rs` and setting [`lib.crate-type` to `staticlib`](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#configuring-a-target) in your Cargo.toml. See also https://os.phil-opp.com/set-up-rust/#creating-a-cargo-project . The resulting `.a` file can then be linked with your C sources, like a normal C staticlib.

**Comment by ghost** [comment]
I know the keyboard input is working
now I have a question how to make the system respond to the keyboard depending on what I write. In the normal rust program it looks like this
```use std::io;
use std::io::prelude::*;
use std::fs::File;

fn main() -> io::Result<()> {
    let mut f = File::open("foo.txt")?;
    let mut buffer = [0; 10];

    // read up to 10 bytes
    f.read(&mut buffer)?;

    let mut buffer = Vec::new();
    // read the whole file
    f.read_to_end(&mut buffer)?;

    // read into a String, so that you don't need to do the conversion.
    let mut buffer = String::new();
    f.read_to_string(&mut buffer)?;

    // and more! See the other methods for more details.
    Ok(())
}```

What does it look like in terms of the operating system?
